include/xtensor/xreducer.hpp

```diff
@@ -34,44 +34,170 @@
+    };
+
+    template <class T, class Tuple>
+    struct tuple_idx_of
     {
-        using type = xtensor<result_type, N - NX, L>;
+        static constexpr std::ptrdiff_t value = tuple_idx_of_impl<0, std::decay_t<T>, typename decay_all<Tuple>::type>::value;
     };
 
-    template <class T, std::size_t... N, layout_type L, class X, std::size_t NX, class result_type>
-    struct xreducer_result_container<xtensor_fixed<T, xshape<N...>, L>, std::array<X, NX>, result_type>
+    template <class T>
+    struct reducer_options
     {
-        using type = xtensor<result_type, sizeof...(N) - NX, L>;
+        reducer_options(T& tpl)
+        {
+            constexpr std::ptrdiff_t initial_val_idx = tuple_idx_of<xinitial<double>, T>::value;
+            if (initial_val_idx != -1)
+            {
+                initial_value = std::get<initial_val_idx>(tpl).value();
+            }
+        }
+
+
+        using evaluation_strategy = std::conditional_t<tuple_idx_of<xt::evaluation_strategy::_immediate, T>::value != -1,
+                                                                    xt::evaluation_strategy::_immediate,
+                                                                    xt::evaluation_strategy::_lazy>;
+
+        using keep_dims = std::conditional_t<tuple_idx_of<xt::_keep_dims, T>::value != -1,
+                                             std::true_type,
+                                             std::false_type>;
+        double initial_value = 0;
     };
 
-    template <class T, std::size_t... I, layout_type L, std::size_t... X, class result_type>
-    struct xreducer_result_container<xtensor_fixed<T, xshape<I...>, L>, xshape<X...>, result_type>
+    template <class T>
+    struct is_reducer_options_impl : std::false_type
     {
-        using type = xtensor_fixed<result_type, typename fixed_xreducer_shape_type<fixed_shape<I...>, fixed_shape<X...>>::type, L>;
     };
 
-    template <class F, class E, class X>
-    inline auto reduce_immediate(F&& f, E&& e, X&& axes)
+    template <class... X>
+    struct is_reducer_options_impl<std::tuple<X...>> : std::true_type
     {
-        using shape_type = typename xreducer_shape_type<typename std::decay_t<E>::shape_type, std::decay_t<X>>::type;
+    };
+
+    template <class T>
+    struct is_reducer_options : is_reducer_options_impl<std::decay_t<T>>
+    {
+    };
+
+    /**********
+     * reduce *
+     **********/
+
+#define DEFAULT_STRATEGY_REDUCERS std::tuple<evaluation_strategy::_lazy>
+
+    template <class ST, class X, class KD = std::false_type>
+    struct xreducer_shape_type;
+
+    template <class S1, class S2>
+    struct fixed_xreducer_shape_type;
+
+    namespace detail
+    {
+        template <class KD, class RS, class R, class E, class AX>
+        inline void shape_computation(RS& result_shape, R& result, E& expr,
+                                      const AX& axes, std::enable_if_t<!detail::is_fixed<RS>::value, int> = 0)
+        {
+            if (KD())
```

Comment:
It's a bit weird to mix `std::false_type` with reducer options unrelated to boolean constants (and rely on their conversion to bool). Besides, this is true for any value of `KD` that do not implement a boolean conversion operator returning false. If the `KD` template parameter cannot be changed for a bool compile-time value, a static constexpr function comparing KD with `keep_dims` option would be more expressive.